File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/viewmodel/HttpCallViewModel.java
package com.prateekj.snooper.networksnooper.viewmodel;

import androidx.annotation.ColorRes;

import com.prateekj.snooper.R;
import com.prateekj.snooper.networksnooper.model.HttpCallRecord;
import com.prateekj.snooper.networksnooper.model.HttpHeader;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.List;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static java.util.Locale.US;

public class HttpCallViewModel {

  private static final String TIMESTAMP_FORMAT = "MM/dd/yyyy HH:mm:ss";
  private static final int RANGE_START_HTTP_OK = 200;
  private static final int RANGE_END_HTTP_OK = 299;
  private static final int RANGE_END_HTTP_REDIRECTION = 399;
  private final HttpCallRecord httpCall;

  public HttpCallViewModel(HttpCallRecord httpCall) {
    this.httpCall = httpCall;
  }

  public String getUrl() {
    return httpCall.getUrl();
  }

  public String getMethod() {
    return httpCall.getMethod();
  }

  public String getStatusCode() {
    return String.valueOf(httpCall.getStatusCode());
  }

  public String getStatusText() {
    return httpCall.getStatusText();
  }

  public List<HttpHeader> getRequestHeaders() {
    return httpCall.getRequestHeaders();
  }

  public List<HttpHeader> getResponseHeaders() {
    return httpCall.getResponseHeaders();
  }

  public String getTimeStamp() {
    DateFormat df = new SimpleDateFormat(TIMESTAMP_FORMAT, US);
    return df.format(httpCall.getDate());
  }

  public @ColorRes int getStatusColor() {
    int statusCode = httpCall.getStatusCode();

    if (statusCode >= RANGE_START_HTTP_OK && statusCode <= RANGE_END_HTTP_OK) {
      return R.color.snooper_green;
    } else if (statusCode <= RANGE_END_HTTP_REDIRECTION) {
      return R.color.snooper_yellow;
    } else {
      return R.color.snooper_red;
    }
  }

  public int getResponseInfoVisibility() {
    return httpCall.hasError() ? GONE : VISIBLE;
  }

  public int getFailedTextVisibility() {
    return httpCall.hasError() ? VISIBLE : GONE;
  }

  public int getResponseHeaderVisibility() {
    return hasHeaders(httpCall.getResponseHeaders()) ? VISIBLE : GONE;
  }

  public int getRequestHeaderVisibility() {
    return hasHeaders(httpCall.getRequestHeaders()) ? VISIBLE : GONE;
  }

  private boolean hasHeaders(List<HttpHeader> headers) {
    return headers != null && !headers.isEmpty();
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/networksnooper/viewmodel/HttpCallViewModel.kt
package com.prateekj.snooper.networksnooper.viewmodel

import android.view.View.GONE
import android.view.View.VISIBLE
import androidx.annotation.ColorRes
import com.prateekj.snooper.R
import com.prateekj.snooper.networksnooper.model.HttpCallRecord
import com.prateekj.snooper.networksnooper.model.HttpHeader
import java.text.SimpleDateFormat
import java.util.Locale.US

class HttpCallViewModel(private val httpCall: HttpCallRecord) {

  val url: String? = httpCall.url

  val method: String? = httpCall.method

  val statusCode: String = httpCall.statusCode.toString()

  val statusText: String? = httpCall.statusText

  val requestHeaders: List<HttpHeader> = httpCall.requestHeaders ?: listOf()

  val responseHeaders: List<HttpHeader> = httpCall.responseHeaders ?: listOf()

  val timeStamp: String
    get() {
      val df = SimpleDateFormat(TIMESTAMP_FORMAT, US)
      return df.format(httpCall.date)
    }

  val responseInfoVisibility: Int = if (httpCall.hasError()) GONE else VISIBLE

  val failedTextVisibility: Int = if (httpCall.hasError()) VISIBLE else GONE

  val responseHeaderVisibility: Int = if (hasHeaders(httpCall.responseHeaders)) VISIBLE else GONE

  val requestHeaderVisibility: Int = if (hasHeaders(httpCall.requestHeaders)) VISIBLE else GONE

  @ColorRes
  fun getStatusColor(): Int {
    val statusCode = httpCall.statusCode
    return when {
      statusCode in RANGE_START_HTTP_OK..RANGE_END_HTTP_OK -> R.color.snooper_green
      statusCode <= RANGE_END_HTTP_REDIRECTION -> R.color.snooper_yellow
      else -> R.color.snooper_red
    }
  }

  private fun hasHeaders(headers: List<HttpHeader>?): Boolean {
    return headers != null && headers.isNotEmpty()
  }

  companion object {
    private const val TIMESTAMP_FORMAT = "MM/dd/yyyy HH:mm:ss"
    private const val RANGE_START_HTTP_OK = 200
    private const val RANGE_END_HTTP_OK = 299
    private const val RANGE_END_HTTP_REDIRECTION = 399
  }
}
